pipelinedp4j/main/com/google/privacy/differentialprivacy/pipelinedp4j/spark/SparkEncoders.kt

```diff
@@ -0,0 +1,198 @@
+//
+//        // Step 2: Define the serializer (Pair<K, V> -> InternalRow)
+//        val serializer = CreateNamedStruct(
+//            JavaConverters.asScalaBuffer(
+//                listOf(keyEncoder.objSerializer(),
+//                    valueEncoder.objSerializer())
+//                ).toSeq())
+//
+//        // Step 3: Define the deserializer (InternalRow -> Pair<K, V>)
+//        val deserializer = CreateNamedStruct(
+//            JavaConverters.asScalaBuffer(
+//                listOf(keyEncoder.objDeserializer(),
+//                    valueEncoder.objDeserializer())
+//            ).toSeq())
+//
+//
+//        val cls1 = keyEncoder.clsTag()
+//        val cls2 = valueEncoder.clsTag()
+//        val pair = Pair(cls1, cls2)
+//        val clsTag = ClassTag.apply<Pair<T1, T2>>(pair.javaClass)
+//
+//        val objSerializer = JavaTypeInference.serializerFor(pair::class.java)
+//        val objDeserializer = JavaTypeInference.deserializerFor(pair::class.java)
+////        val d = JavaTypeInference.inferDataType(fir)
+//
+//        // Step 4: Combine into an ExpressionEncoder
+//        return ExpressionEncoder(
+//            serializer,
+//            PairDeserializer<T1, T2>(keyEncoder.objDeserializer(), valueEncoder.objDeserializer()),
+//            clsTag // scala.reflect.classTag<Pair<K, V>>() // Ensures runtime type safety
+//        )
+//    }
+//}
+//
+//class PairDeserializer<T1, T2>(
+//    private val keyDeserializer: Expression,
+//    private val valueDeserializer: Expression
+//) : Expression() {
+//
+//    override fun nullable(): Boolean = false
+//
+//    override fun dataType(): StructType = StructType(
+//        listOf(
+//            StructField("key", keyDeserializer.dataType(), keyDeserializer.nullable(), Metadata.empty()),
+//            StructField("value", valueDeserializer.dataType(), valueDeserializer.nullable(), Metadata.empty())
+//        ).toTypedArray()
+//    )
+//
+//    override fun eval(input: InternalRow?): Pair<T1, T2>? {
+//        if (input == null) return null
+//        val key = keyDeserializer.eval(input) as T1
+//        val value = valueDeserializer.eval(input) as T2
+//        return Pair(key, value)
+//    }
+//
+//    override fun canEqual(that: Any?): Boolean {
+//        TODO("Not yet implemented")
+//    }
+//
+//    override fun productElement(n: Int): Any {
+//        TODO("Not yet implemented")
+//    }
+//
+//    override fun productArity(): Int {
+//        TODO("Not yet implemented")
+//    }
+//
+//    override fun children(): Seq<Expression> =
+//        JavaConverters.asScalaBuffer(listOf(keyDeserializer, valueDeserializer)).toSeq()
+//
+//    override fun withNewChildrenInternal(newChildren: IndexedSeq<Expression>?): Expression {
+//        TODO("Not yet implemented")
+//    }
+//
+//    override fun doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode {
+//        throw UnsupportedOperationException("Code generation is not supported for PairDeserializer")
+//    }
+//}
```

Comment:
yes it was complicated because was working in internal spark lib. Replaced with `kryo` serializer deserializer for ow which seems to be doing what we want but not very efficient since `kryo` is not native to spark `catalyst optimizer`.